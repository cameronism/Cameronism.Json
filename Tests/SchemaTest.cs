using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using Xunit;

namespace Tests
{
	public class SchemaTest
	{
		#region converters
		class TypeConverter : Newtonsoft.Json.JsonConverter
		{
			public override bool CanConvert(Type objectType)
			{
				return typeof(Type).IsAssignableFrom(objectType);
			}

			public override object ReadJson(Newtonsoft.Json.JsonReader reader, Type objectType, object existingValue, Newtonsoft.Json.JsonSerializer serializer)
			{
				throw new NotSupportedException();
			}

			public override void WriteJson(Newtonsoft.Json.JsonWriter writer, object value, Newtonsoft.Json.JsonSerializer serializer)
			{
				if (value == null)
				{
					writer.WriteNull();
					return;
				}

				writer.WriteValue(HumanName((Type)value));
			}
		}

		class MemberInfoConverter : Newtonsoft.Json.JsonConverter
		{
			public override bool CanConvert(Type objectType)
			{
				return typeof(FieldInfo).IsAssignableFrom(objectType) || typeof(PropertyInfo).IsAssignableFrom(objectType);
			}

			public override object ReadJson(Newtonsoft.Json.JsonReader reader, Type objectType, object existingValue, Newtonsoft.Json.JsonSerializer serializer)
			{
				throw new NotSupportedException();
			}

			public override void WriteJson(Newtonsoft.Json.JsonWriter writer, object value, Newtonsoft.Json.JsonSerializer serializer)
			{
				var fi = value as FieldInfo;
				var pi = value as PropertyInfo;
				var details = new
				{
					Name = 
						fi != null ? fi.Name : 
						pi != null ? pi.Name :
						null,
					MemberType =
						fi != null ? fi.FieldType : 
						pi != null ? pi.PropertyType :
						null,
				};

				if (details.Name != null)
				{
					serializer.Serialize(writer, details);
				}
				else
				{
					writer.WriteNull();
				}
			}
		}
		#endregion

		[System.Runtime.Serialization.DataContract]
		class ExplicitDataMemberOrder
		{
			[System.Runtime.Serialization.DataMember(Order=3)]
			public int C;
			[System.Runtime.Serialization.DataMember(Order=2)]
			public int B;
			[System.Runtime.Serialization.DataMember(Order=1)]
			public int A;
		}

		public static string HumanName(Type t)
		{
			var ns = t.Namespace;
			var name = t.Name;
			if (t.IsGenericType)
			{
				name = name.Substring(0, name.IndexOf('`')) + "<" +  String.Join(", ", t.GetGenericArguments().Select(a => HumanName(a))) + ">";

			}
			return String.IsNullOrEmpty(ns) ? name : ns + "." + name;
		}

		static Type GetType<T>(T sample)
		{
			return typeof(T);
		}

		[Fact]
		public void Reflect()
		{
			var converters = new Newtonsoft.Json.JsonConverter[] { 
				new TypeConverter(), 
				new Newtonsoft.Json.Converters.StringEnumConverter { },
				new MemberInfoConverter(),
			};
			var sb = new StringBuilder();

			foreach (var t in new[] { 
				typeof(System.Int32),
				typeof(System.Boolean),
				typeof(System.Byte),
				typeof(System.Char),
				typeof(System.DateTime),
				typeof(System.Decimal),
				typeof(System.Double),
				typeof(System.Int16),
				typeof(System.Int64),
				typeof(System.SByte),
				typeof(System.Single),
				typeof(System.String),
				typeof(System.UInt16),
				typeof(System.UInt32),
				typeof(System.UInt64),

				typeof(System.Int32?),
				typeof(System.Boolean?),
				typeof(System.Byte?),
				typeof(System.Char?),
				typeof(System.DateTime?),
				typeof(System.Decimal?),
				typeof(System.Double?),
				typeof(System.Int16?),
				typeof(System.Int64?),
				typeof(System.SByte?),
				typeof(System.Single?),
				typeof(System.UInt16?),
				typeof(System.UInt32?),
				typeof(System.UInt64?),

				typeof(System.Guid),
				typeof(System.Guid?),

				typeof(IEnumerable<int>),
				typeof(string[]),
				typeof(IReadOnlyList<Guid>),

				GetType(new { i = 1 }),
				GetType(new[] { new { s = "" } }.ToList()),

				typeof(Dictionary<string, string>),
				typeof(IReadOnlyDictionary<Guid, System.DayOfWeek>),

				typeof(KeyValuePair<int, int>),
				typeof(KeyValuePair<int, int>?),

				typeof(DateTimeOffset),
				typeof(DateTimeOffset?),

				typeof(System.Net.IPAddress),

				typeof(ExplicitDataMemberOrder),
			})
			{
				var schema = UnsafeJson.Schema.Reflect(t);
				var json = Newtonsoft.Json.JsonConvert.SerializeObject(schema, Newtonsoft.Json.Formatting.Indented, converters);

				sb.AppendLine();
				sb.AppendLine();
				sb.AppendLine("# " + HumanName(t));
				sb.AppendLine(json);
			}


			ApprovalTests.Approvals.Verify(sb.ToString());
		}
	}
}
