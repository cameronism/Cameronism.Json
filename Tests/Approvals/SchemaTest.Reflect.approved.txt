

# System.Int32
{
  "NetType": "System.Int32",
  "JsonType": "Integer",
  "Nullable": false,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Boolean
{
  "NetType": "System.Boolean",
  "JsonType": "Boolean",
  "Nullable": false,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Byte
{
  "NetType": "System.Byte",
  "JsonType": "Integer",
  "Nullable": false,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Char
{
  "NetType": "System.Char",
  "JsonType": "String",
  "Nullable": false,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.DateTime
{
  "NetType": "System.DateTime",
  "JsonType": "String",
  "Nullable": false,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Decimal
{
  "NetType": "System.Decimal",
  "JsonType": "Number",
  "Nullable": false,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Double
{
  "NetType": "System.Double",
  "JsonType": "Number",
  "Nullable": false,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Int16
{
  "NetType": "System.Int16",
  "JsonType": "Integer",
  "Nullable": false,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Int64
{
  "NetType": "System.Int64",
  "JsonType": "Integer",
  "Nullable": false,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.SByte
{
  "NetType": "System.SByte",
  "JsonType": "Integer",
  "Nullable": false,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Single
{
  "NetType": "System.Single",
  "JsonType": "Number",
  "Nullable": false,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.String
{
  "NetType": "System.String",
  "JsonType": "String",
  "Nullable": true,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.UInt16
{
  "NetType": "System.UInt16",
  "JsonType": "Integer",
  "Nullable": false,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.UInt32
{
  "NetType": "System.UInt32",
  "JsonType": "Integer",
  "Nullable": false,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.UInt64
{
  "NetType": "System.UInt64",
  "JsonType": "Integer",
  "Nullable": false,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Nullable<System.Int32>
{
  "NetType": "System.Nullable<System.Int32>",
  "JsonType": "Integer",
  "Nullable": true,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Nullable<System.Boolean>
{
  "NetType": "System.Nullable<System.Boolean>",
  "JsonType": "Boolean",
  "Nullable": true,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Nullable<System.Byte>
{
  "NetType": "System.Nullable<System.Byte>",
  "JsonType": "Integer",
  "Nullable": true,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Nullable<System.Char>
{
  "NetType": "System.Nullable<System.Char>",
  "JsonType": "String",
  "Nullable": true,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Nullable<System.DateTime>
{
  "NetType": "System.Nullable<System.DateTime>",
  "JsonType": "String",
  "Nullable": true,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Nullable<System.Decimal>
{
  "NetType": "System.Nullable<System.Decimal>",
  "JsonType": "Number",
  "Nullable": true,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Nullable<System.Double>
{
  "NetType": "System.Nullable<System.Double>",
  "JsonType": "Number",
  "Nullable": true,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Nullable<System.Int16>
{
  "NetType": "System.Nullable<System.Int16>",
  "JsonType": "Integer",
  "Nullable": true,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Nullable<System.Int64>
{
  "NetType": "System.Nullable<System.Int64>",
  "JsonType": "Integer",
  "Nullable": true,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Nullable<System.SByte>
{
  "NetType": "System.Nullable<System.SByte>",
  "JsonType": "Integer",
  "Nullable": true,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Nullable<System.Single>
{
  "NetType": "System.Nullable<System.Single>",
  "JsonType": "Number",
  "Nullable": true,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Nullable<System.UInt16>
{
  "NetType": "System.Nullable<System.UInt16>",
  "JsonType": "Integer",
  "Nullable": true,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Nullable<System.UInt32>
{
  "NetType": "System.Nullable<System.UInt32>",
  "JsonType": "Integer",
  "Nullable": true,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Nullable<System.UInt64>
{
  "NetType": "System.Nullable<System.UInt64>",
  "JsonType": "Integer",
  "Nullable": true,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Guid
{
  "NetType": "System.Guid",
  "JsonType": "String",
  "Nullable": false,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Nullable<System.Guid>
{
  "NetType": "System.Nullable<System.Guid>",
  "JsonType": "String",
  "Nullable": true,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Collections.Generic.IEnumerable<System.Int32>
{
  "NetType": "System.Collections.Generic.IEnumerable<System.Int32>",
  "JsonType": "Array",
  "Nullable": true,
  "Items": {
    "NetType": "System.Int32",
    "JsonType": "Integer",
    "Nullable": false,
    "Items": null,
    "Keys": null,
    "Members": null,
    "FieldInfo": null,
    "PropertyInfo": null
  },
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.String[]
{
  "NetType": "System.String[]",
  "JsonType": "Array",
  "Nullable": true,
  "Items": {
    "NetType": "System.String",
    "JsonType": "String",
    "Nullable": true,
    "Items": null,
    "Keys": null,
    "Members": null,
    "FieldInfo": null,
    "PropertyInfo": null
  },
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Collections.Generic.IReadOnlyList<System.Guid>
{
  "NetType": "System.Collections.Generic.IReadOnlyList<System.Guid>",
  "JsonType": "Array",
  "Nullable": true,
  "Items": {
    "NetType": "System.Guid",
    "JsonType": "String",
    "Nullable": false,
    "Items": null,
    "Keys": null,
    "Members": null,
    "FieldInfo": null,
    "PropertyInfo": null
  },
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# <>f__AnonymousType1<System.Int32>
{
  "NetType": "<>f__AnonymousType1<System.Int32>",
  "JsonType": "Object",
  "Nullable": true,
  "Items": null,
  "Keys": null,
  "Members": [
    {
      "Key": "i",
      "Value": {
        "NetType": "System.Int32",
        "JsonType": "Integer",
        "Nullable": false,
        "Items": null,
        "Keys": null,
        "Members": null,
        "FieldInfo": null,
        "PropertyInfo": {
          "Name": "i",
          "MemberType": "System.Int32"
        }
      }
    }
  ],
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Collections.Generic.List<<>f__AnonymousType2<System.String>>
{
  "NetType": "System.Collections.Generic.List<<>f__AnonymousType2<System.String>>",
  "JsonType": "Array",
  "Nullable": true,
  "Items": {
    "NetType": "<>f__AnonymousType2<System.String>",
    "JsonType": "Object",
    "Nullable": true,
    "Items": null,
    "Keys": null,
    "Members": [
      {
        "Key": "s",
        "Value": {
          "NetType": "System.String",
          "JsonType": "String",
          "Nullable": true,
          "Items": null,
          "Keys": null,
          "Members": null,
          "FieldInfo": null,
          "PropertyInfo": {
            "Name": "s",
            "MemberType": "System.String"
          }
        }
      }
    ],
    "FieldInfo": null,
    "PropertyInfo": null
  },
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Collections.Generic.Dictionary<System.String, System.String>
{
  "NetType": "System.Collections.Generic.Dictionary<System.String, System.String>",
  "JsonType": "Object",
  "Nullable": true,
  "Items": {
    "NetType": "System.String",
    "JsonType": "String",
    "Nullable": true,
    "Items": null,
    "Keys": null,
    "Members": null,
    "FieldInfo": null,
    "PropertyInfo": null
  },
  "Keys": {
    "NetType": "System.String",
    "JsonType": "String",
    "Nullable": true,
    "Items": null,
    "Keys": null,
    "Members": null,
    "FieldInfo": null,
    "PropertyInfo": null
  },
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Collections.Generic.IReadOnlyDictionary<System.Guid, System.DayOfWeek>
{
  "NetType": "System.Collections.Generic.IReadOnlyDictionary<System.Guid, System.DayOfWeek>",
  "JsonType": "Object",
  "Nullable": true,
  "Items": {
    "NetType": "System.DayOfWeek",
    "JsonType": "Integer",
    "Nullable": false,
    "Items": null,
    "Keys": null,
    "Members": null,
    "FieldInfo": null,
    "PropertyInfo": null
  },
  "Keys": {
    "NetType": "System.Guid",
    "JsonType": "String",
    "Nullable": false,
    "Items": null,
    "Keys": null,
    "Members": null,
    "FieldInfo": null,
    "PropertyInfo": null
  },
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Collections.Generic.KeyValuePair<System.Int32, System.Int32>
{
  "NetType": "System.Collections.Generic.KeyValuePair<System.Int32, System.Int32>",
  "JsonType": "Object",
  "Nullable": false,
  "Items": null,
  "Keys": null,
  "Members": [
    {
      "Key": "Key",
      "Value": {
        "NetType": "System.Int32",
        "JsonType": "Integer",
        "Nullable": false,
        "Items": null,
        "Keys": null,
        "Members": null,
        "FieldInfo": null,
        "PropertyInfo": {
          "Name": "Key",
          "MemberType": "System.Int32"
        }
      }
    },
    {
      "Key": "Value",
      "Value": {
        "NetType": "System.Int32",
        "JsonType": "Integer",
        "Nullable": false,
        "Items": null,
        "Keys": null,
        "Members": null,
        "FieldInfo": null,
        "PropertyInfo": {
          "Name": "Value",
          "MemberType": "System.Int32"
        }
      }
    }
  ],
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Nullable<System.Collections.Generic.KeyValuePair<System.Int32, System.Int32>>
{
  "NetType": "System.Nullable<System.Collections.Generic.KeyValuePair<System.Int32, System.Int32>>",
  "JsonType": "Object",
  "Nullable": true,
  "Items": null,
  "Keys": null,
  "Members": [
    {
      "Key": "Key",
      "Value": {
        "NetType": "System.Int32",
        "JsonType": "Integer",
        "Nullable": false,
        "Items": null,
        "Keys": null,
        "Members": null,
        "FieldInfo": null,
        "PropertyInfo": {
          "Name": "Key",
          "MemberType": "System.Int32"
        }
      }
    },
    {
      "Key": "Value",
      "Value": {
        "NetType": "System.Int32",
        "JsonType": "Integer",
        "Nullable": false,
        "Items": null,
        "Keys": null,
        "Members": null,
        "FieldInfo": null,
        "PropertyInfo": {
          "Name": "Value",
          "MemberType": "System.Int32"
        }
      }
    }
  ],
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.DateTimeOffset
{
  "NetType": "System.DateTimeOffset",
  "JsonType": "String",
  "Nullable": false,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Nullable<System.DateTimeOffset>
{
  "NetType": "System.Nullable<System.DateTimeOffset>",
  "JsonType": "String",
  "Nullable": true,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# System.Net.IPAddress
{
  "NetType": "System.Net.IPAddress",
  "JsonType": "String",
  "Nullable": true,
  "Items": null,
  "Keys": null,
  "Members": null,
  "FieldInfo": null,
  "PropertyInfo": null
}


# Tests.SchemaTest+ExplicitDataMemberOrder
{
  "NetType": "Tests.SchemaTest+ExplicitDataMemberOrder",
  "JsonType": "Object",
  "Nullable": true,
  "Items": null,
  "Keys": null,
  "Members": [
    {
      "Key": "A",
      "Value": {
        "NetType": "System.Int32",
        "JsonType": "Integer",
        "Nullable": false,
        "Items": null,
        "Keys": null,
        "Members": null,
        "FieldInfo": {
          "Name": "A",
          "MemberType": "System.Int32"
        },
        "PropertyInfo": null
      }
    },
    {
      "Key": "B",
      "Value": {
        "NetType": "System.Int32",
        "JsonType": "Integer",
        "Nullable": false,
        "Items": null,
        "Keys": null,
        "Members": null,
        "FieldInfo": {
          "Name": "B",
          "MemberType": "System.Int32"
        },
        "PropertyInfo": null
      }
    },
    {
      "Key": "C",
      "Value": {
        "NetType": "System.Int32",
        "JsonType": "Integer",
        "Nullable": false,
        "Items": null,
        "Keys": null,
        "Members": null,
        "FieldInfo": {
          "Name": "C",
          "MemberType": "System.Int32"
        },
        "PropertyInfo": null
      }
    }
  ],
  "FieldInfo": null,
  "PropertyInfo": null
}
